docs/proposals/federated-ruler.md

```diff
@@ -0,0 +1,84 @@
+
+Additionally an organization may wish to treat several cortex tenants `t0|t1|...|ti` as one logical tenant due to compactor scalability. In this case a composite tenant `t0|t1|...|ti` would own a federated rule, which queries each subtenant `t0` thru `ti`. The resulting data could be sent to a specific tenant e.g. `admin` or as the composite tenant treats each subtenant uniformly, a random, but consistent subtenant can be chosen to store the resulting series in, in this case `tj` where 0 <= j <= i.
+More explicitly, for a given recording rule and the produced series `foobarbaz` which is owned by the composite tenant `0|1|2|3`, a subtenant `0`, `1`, `2` or `3` is chosen to always store the series `foobarbaz` in, lets say `2`. Another rule and produced series `fizzbuzz` owned by the same composite tenant `0|1|2|3` makes this same chose of subtenant, in our case chosing `0`.
+
+## Challenges
+
+### Allow federated rules behind feature flag
+
+#### Challenge
+
+Federated tenant rules and alerts will not be a good fit for organization and should be behind a feature flag.
+
+#### Proposal
+
+For federated rules owned by a single tenant, creation of federated rules (those sourcing data from multiple tenants) should be blocked behind the feature flag `tenant-federation.enabled`.
+
+To support composite tenants, if tenant federation is enabled for ruler and alertmanager via `tenant-federation.enabled`, then ruler use a `mergeQueryable` to aggregate the results of querying multiple tenants. The ruler and alertmanager APIs should be updated to always call `tenant.GetTenantIDs` instead of `tenant.GetTenantID`, which will use the MultiTenantResolver when tenant federation is enabled.
+
+### Allow federated rules only for select tenants
+
+#### Challenge
+
+For many organizations, the ability for any tenant to write a rule querying any other tenant is not acceptable and more fine grained control is required
+
+#### Proposal
+
+Since the current default is that a tenant should only be able to write rules against itself, we suggest a config option `tenant-federation.ruler-federated-tenant-allow-list`, a string slice of OrgIDs like `infra` or `0|1|2|3|4` which are allowed to write rules against all tenants. If a tenant `bar` attempts to create a federated rule, an error should be returned by the ruler api. If a tenant `bar` contains a federated rule in the rule store, but is not in the allow list, an error should be logged and the federated rule should not be evaluated.
+
+If more fine grained access control is required, we suggest implementing a similar config option `tenant-federation.ruler-federated-tenant-allow-map` containing a json string -> string map of OrgID -> Permitted OrgIDs e.g. a value of `{"infra":"0|1|2|3|4", "0|1|2":"0|1|2"}` should allow `infra` to query any combination of the tenants `0`, `1`, `2`, `3`, and `4`. Behavior for creation and execution of not permitted federated rules should be similar to the above.
```

Comment:
To match current behavior I'm implementing 
```
      AllowedFederatedTenants    flagext.StringSliceCSV `yaml:"allowed_federated_tenants"`
     DisallowedFederatedTenants flagext.StringSliceCSV `yaml:"disallowed_federated_tenants"`
```
Which are checked along with `ruler.enable_federated_rules` at rule creation